### ISSUE TITLE
Kernel assertion: USB code allocates memory region with size of 0 bytes

### ISSUE THREAD

**Issue by JayFoxRox** [XboxDev/nxdk#121]
In a private project I wrote a custom USB driver, and when running it on a debug Xbox it triggered a kernel assertion for `MmAllocateContiguousMemoryEx` with size 0.

I'm not sure where exactly this originates yet, but the function which does these allocations is:

https://github.com/XboxDev/nxdk/blob/82e4860c2f5a358677f60000e40f8f00529e5e8e/lib/usb/misc/malloc.c#L22-L30

A temporary workaround is this to add `if (s == 0) { s = 1; }`.

For debugging this issue (finding responsible caller), someone should turn `malloc2` into a macro, so it reports `__FILE__` and `__LINE__`; or use `RtlCaptureStackBackTrace` (although I'm not sure how to use this).

**Comment by mborgerson** [comment]
If your software can be used in XQEMU (which I assume it can), one can also easily debug this using gdb, load the symbol file, and collect a backtrace.

**Comment by JayFoxRox** [comment]
Thanks to @kl0wn , who debugged this with gdb. He was unable to reproduce it in the Oldskool sample.
So I ended up `printf`-debugging it myself and was also unable to reproduce it in Oldskool.

I then moved back to the original project where it had caused trouble. This confirmed it's a bug in nxdk, that simply isn't triggered by existing drivers.

In particular it's here: https://github.com/XboxDev/nxdk/blob/b3c31d84cb4786acd5b6ed45012839aaa4d66903/lib/usb/core/config.c#L217-L225

`ifp->desc.bNumEndpoints` is zero, for the Xbox 360 gamepad that I'm using.

---

The [linux kernel solves it by having a dummy pointer](https://github.com/torvalds/linux/blob/01ccc3ad44130458769646204449e2e4124f15da/include/linux/slab.h#L118-L129).
This solution [has existed for at least 12 years](https://github.com/torvalds/linux/commit/6cb8f91320d3e720351c21741da795fed580b21b). However, I did not check if the linux kernel might have had a different solution when our USB code was written (probably 15 years ago).

Someone should probably PR a similar approach for nxdk.
